snuba/web/query.py

```diff
@@ -119,15 +124,22 @@ def parse_and_run_query(
             robust=robust,
             concurrent_queries_gauge=concurrent_queries_gauge,
         )
+        _set_query_final(request, result.extra)
         if not request.settings.get_dry_run():
             record_query(request, timer, query_metadata, result.extra)
     except QueryException as error:
+        _set_query_final(request, error.extra)
         record_query(request, timer, query_metadata, error.extra)
         raise error
 
     return result
 
 
+def _set_query_final(request: Request, extra: QueryExtraData) -> None:
+    if "final" in extra["stats"]:
```

Comment:
Not sure if this check is necessary but `QueryExtraData.stats` isn't a `TypedDict` so I'm not sure the `"final"` key always exists.